### PR TITLE
Coerce timestamps to the supported precision

### PR DESCRIPTION
**Pull request by tjwp**
With this change, instances match exactly after a round-trip through serialization and deserialization.

Previously a timestamp stored in a model could have greater precision than a value returned after deserialization.

This removes any need to truncate timestamps before assigning to an Avromatic model.

Prime: @jturkel

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,8 +7,7 @@
 - Add support for recursive models.
 - Allow required array and map fields to be empty. Only nil values for required
   array and map fields are now considered invalid.
-- Validate nested models. This does not apply to nested models embedded
```
Comment (jturkel) [inline CHANGELOG.md]: Merge error?

File: lib/avromatic/model/attribute/abstract_timestamp.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      # to the supported precision.
+      class AbstractTimestamp < Virtus::Attribute
+        def coerce(value)
+          return value if value.nil? || value_coerced?(value)
+
+          value.is_a?(Time) ? coerce_time(value) : value
+        end
+
+        def value_coerced?(_value)
+          raise 'subclass must implement `value_coerced?`'
```
Comment (jturkel) [inline lib/avromatic/model/attribute/abstract_timestamp.rb]: We would get some more useful context if the message was `"#{__method__} must be overridden by #{self.class.name}"`

File: lib/avromatic/model/attribute/timestamp_micros.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      class TimestampMicros < AbstractTimestamp
+
+        def value_coerced?(value)
+          value.is_a?(Time) && value.nsec % 1000 == 0
+        end
+
+        private
+
+        def coerce_time(value)
+          Time.at(value.to_i, value.usec)
```
Comment (jturkel) [inline lib/avromatic/model/attribute/timestamp_micros.rb]: How does this work with timezones?

File: lib/avromatic/model/attribute/timestamp_millis.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      class TimestampMillis < AbstractTimestamp
+
+        def value_coerced?(value)
+          value.is_a?(Time) && value.usec % 1000 == 0
+        end
+
+        private
+
+        def coerce_time(value)
+          Time.at(value.to_i, value.usec / 1000 * 1000)
```
Comment (jturkel) [inline lib/avromatic/model/attribute/timestamp_millis.rb]: Same question about timezones.

File: CHANGELOG.md (hunk 1 above)
Comment (tjwp) [inline CHANGELOG.md]: I think this was prior to me merging with master.

File: lib/avromatic/model/attribute/timestamp_micros.rb (hunk 3 above)
Comment (tjwp) [inline lib/avromatic/model/attribute/timestamp_micros.rb]: The `Time` value created here is local time. The representation as a number of seconds is always Epoch (UTC).

**Comment by jturkel** [comment]
Looks good. :shipit:

**Pull request merged**